### ISSUE TITLE
[BUG] `Differencer` in `ForecastingPipeline` and `TransformedTargetForecaster` does not work!

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/sktime#2452]
### Discussed in https://github.com/alan-turing-institute/sktime/discussions/2451

<div type='discussions-op-text'>

<sup>Originally posted by **anthonygiorgio97** April 13, 2022</sup>
I was trying to do a Pipeline with transformers for both exogenous X (4 features) and endogenous y variable with a FeatureSelection(), a LogTransformer() and a Differencer() combined with a cross validation to tune Prophet (but the same happen with other algorithms). The difference operator must exclude the first index from both exogenous and endogenous variable (because I have specified this transformation both in ForecastingPipeline and TransformedTargetForecaster), but when I call the fit method this error raise: 

(time) indices of series 0 are not contained in index of series 0, extra indices are: DatetimeIndex(['2016-01-31'], dtype='datetime64[ns]', freq=None)

where 2016-01-31 is the first index. If I use a classic Detrender() the code work, but I want to try with the Differencer() operator.

I am using sktime 0.11.0. This is the code I am using:

```python
pipe_prophet = ForecastingPipeline(steps=[
        ("selection", FeatureSelection(method = 'columns')),
        ("logX", OptionalPassthrough(LogTransformer())),
        ("differencerX", Differencer()),
        ("prophetforecaster", TransformedTargetForecaster(steps = [
            ("log", OptionalPassthrough(LogTransformer())),
            ("differencer", Differencer()),
            ("prophet", Prophet(yearly_seasonality='auto'))
        ]))  
])

cv = ExpandingWindowSplitter(initial_window = len(train_model) - (N_cv_fold - 1) * step_cv - len(fh),
                            start_with_window = True,
                            step_length = step_cv,
                            fh = list(range(1,forecasting_horizon +1,1)))

param_grid = [{"logX__passthrough": [False],
               "prophetforecaster__log__passthrough": [False],
               "selection__columns": [[prod_group + ' ADJ'],[prod_group + ' ADJ','MONTHLY_ROLLING_MEAN'],[prod_group + ' ADJ','MONTHLY_ROLLING_MEAN','MONTHLY_ROLLING_MIN','MONTHLY_ROLLING_MAX']],
               "prophetforecaster__prophet__seasonality_mode": ['additive','multiplicative'],
               "prophetforecaster__prophet__add_country_holidays": [None, {'country_name': 'Italy'}]},
              {"logX__passthrough": [True],
               "prophetforecaster__log__passthrough": [True],
               "selection__columns": [[prod_group + ' ADJ'],[prod_group + ' ADJ','MONTHLY_ROLLING_MEAN'],[prod_group + ' ADJ','MONTHLY_ROLLING_MEAN','MONTHLY_ROLLING_MIN','MONTHLY_ROLLING_MAX']],
               "prophetforecaster__prophet__seasonality_mode": ['additive','multiplicative'],
               "prophetforecaster__prophet__add_country_holidays": [None, {'country_name': 'Italy'}]}]

gscv_prophet = ForecastingGridSearchCV(
            forecaster = pipe_prophet,
            cv=cv,
            param_grid=param_grid,
            verbose = 1
        )
gscv_prophet.fit(train_model, X = df_ex_train, fh = fh)
```

</div>

**Comment by fkiraly** [comment]
@anthonygiorgio97 to help us fix this quicker, would you mind posting full code, incl imports, replacing your data by some example data? You can get example data from `sktime.datatypes.get_examples`, or `utils._testing.series._make_series`.

**Comment by fkiraly** [comment]
Cannot reproduce this with an attempted replication:

```python
from sktime.forecasting.ets import AutoETS
from sktime.transformations.series.boxcox import LogTransformer
from sktime.transformations.series.compose import OptionalPassthrough
from sktime.transformations.series.difference import Differencer
from sktime.utils._testing.series import _make_series

y = _make_series()
pipe = ForecastingPipeline(steps=[
        ("logX", OptionalPassthrough(LogTransformer())),
        ("differencerX", Differencer()),
        ("prophetforecaster", TransformedTargetForecaster(steps = [
            ("log", OptionalPassthrough(LogTransformer())),
            ("differencer", Differencer()),
            ("prophet", AutoETS())
        ]))  
])
pipe.fit(y, fh=[1, 2, 3]).predict()
```

**Comment by fkiraly** [comment]
@anthonygiorgio97, can you please post a self-contained code snippet?

**Comment by anthonygiorgio97** [comment]
This is a code example, for version 0.11.0:

```
from sktime.transformations.series.boxcox import LogTransformer, BoxCoxTransformer
from sktime.forecasting.compose import ForecastingPipeline, TransformedTargetForecaster
from sktime.transformations.series.compose import OptionalPassthrough
from sktime.transformations.series.feature_selection import FeatureSelection
from sktime.forecasting.fbprophet import Prophet
from sktime.transformations.series.difference import Differencer
from sktime.forecasting.model_selection import ExpandingWindowSplitter, temporal_train_test_split, ForecastingGridSearchCV
from sktime.datasets import load_longley
y, X = load_longley()

#split train/test both y and X
fh = [1,2,3]
train_model, test_model = temporal_train_test_split(y, fh = fh)
X_train = X[X.index.isin(train_model.index)]
X_test = X[X.index.isin(test_model.index)]

#prophet pipeline
pipe_prophet = ForecastingPipeline(steps=[
        ("boxcoxX", OptionalPassthrough(BoxCoxTransformer())),
        ("differencerX", Differencer(drop_na = True)),
        ("prophetforecaster", TransformedTargetForecaster(steps = [
            ("boxcox", OptionalPassthrough(BoxCoxTransformer())),
            ("differencer", Differencer(drop_na = True)),
            ("prophet", Prophet(yearly_seasonality='auto'))
        ]))  
])

#cv setup
N_cv_fold = 2
step_cv = 3
cv = ExpandingWindowSplitter(initial_window = len(train_model) - (N_cv_fold - 1) * step_cv - len(fh),
                            start_with_window = True,
                            step_length = step_cv,
                            fh = fh)

param_grid = [{"boxcoxX__passthrough": [False],
               "prophetforecaster__boxcox__passthrough": [False],
               "prophetforecaster__prophet__seasonality_mode": ['additive','multiplicative']},
              {"boxcoxX__passthrough": [True],
               "prophetforecaster__boxcox__passthrough": [True],
               "prophetforecaster__prophet__seasonality_mode": ['additive','multiplicative']}
             ]

#grid search
gscv_prophet = ForecastingGridSearchCV(
            forecaster = pipe_prophet,
            cv=cv,
            param_grid=param_grid,
            verbose = 1
        )

#fit
gscv_prophet.fit(train_model, X = X_train, fh = [1,2,3])
```

This error raise:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<[command-351381282254798]()> in <module>
     51 
     52 #fit
---> 53 gscv_prophet.fit(train_model, X = X_train, fh = [1,2,3])

/databricks/python/lib/python3.8/site-packages/sktime/forecasting/base/_base.py in fit(self, y, X, fh)
    161         # we call the ordinary _fit if no looping/vectorization needed
    162         if not vectorization_needed:
--> 163             self._fit(y=y_inner, X=X_inner, fh=fh)
    164         else:
    165             # otherwise we call the vectorized version of fit

/databricks/python/lib/python3.8/site-packages/sktime/forecasting/model_selection/_tune.py in _fit(self, y, X, fh, **fit_params)
    255 
    256         # Run grid-search cross-validation.
--> 257         results = self._run_search(evaluate_candidates)
    258 
    259         results = pd.DataFrame(results)

/databricks/python/lib/python3.8/site-packages/sktime/forecasting/model_selection/_tune.py in _run_search(self, evaluate_candidates)
    463         """Search all candidates in param_grid."""
    464         _check_param_grid(self.param_grid)
--> 465         return evaluate_candidates(ParameterGrid(self.param_grid))
    466 
    467     @classmethod

/databricks/python/lib/python3.8/site-packages/sktime/forecasting/model_selection/_tune.py in evaluate_candidates(candidate_params)
    241                 )
    242 
--> 243             out = parallel(
    244                 delayed(_fit_and_score)(params) for params in candidate_params
    245             )

/databricks/python/lib/python3.8/site-packages/joblib/parallel.py in __call__(self, iterable)
   1039             # remaining jobs.
   1040             self._iterating = False
-> 1041             if self.dispatch_one_batch(iterator):
   1042                 self._iterating = self._original_iterator is not None
   1043 

/databricks/python/lib/python3.8/site-packages/joblib/parallel.py in dispatch_one_batch(self, iterator)
    857                 return False
    858             else:
--> 859                 self._dispatch(tasks)
    860                 return True
    861 

/databricks/python/lib/python3.8/site-packages/joblib/parallel.py in _dispatch(self, batch)
    775         with self._lock:
    776             job_idx = len(self._jobs)
--> 777             job = self._backend.apply_async(batch, callback=cb)
    778             # A job can complete so quickly than its callback is
    779             # called before we get here, causing self._jobs to

/databricks/python/lib/python3.8/site-packages/joblib/_parallel_backends.py in apply_async(self, func, callback)
    206     def apply_async(self, func, callback=None):
    207         """Schedule a func to be run"""
--> 208         result = ImmediateResult(func)
    209         if callback:
    210             callback(result)

/databricks/python/lib/python3.8/site-packages/joblib/_parallel_backends.py in __init__(self, batch)
    570         # Don't delay the application, to avoid keeping the input
    571         # arguments in memory
--> 572         self.results = batch()
    573 
    574     def get(self):

/databricks/python/lib/python3.8/site-packages/joblib/parallel.py in __call__(self)
    260         # change the default number of processes to -1
    261         with parallel_backend(self._backend, n_jobs=self._n_jobs):
--> 262             return [func(*args, **kwargs)
    263                     for func, args, kwargs in self.items]
    264 

/databricks/python/lib/python3.8/site-packages/joblib/parallel.py in <listcomp>(.0)
    260         # change the default number of processes to -1
    261         with parallel_backend(self._backend, n_jobs=self._n_jobs):
--> 262             return [func(*args, **kwargs)
    263                     for func, args, kwargs in self.items]
    264 

/databricks/python/lib/python3.8/site-packages/sktime/forecasting/model_selection/_tune.py in _fit_and_score(params)
    206 
    207             # Evaluate.
--> 208             out = evaluate(
    209                 forecaster,
    210                 cv,

/databricks/python/lib/python3.8/site-packages/sktime/forecasting/model_evaluation/_functions.py in evaluate(forecaster, cv, y, X, strategy, scoring, fit_params, return_data)
    108         if i == 0 or strategy == "refit":
    109             forecaster = clone(forecaster)
--> 110             forecaster.fit(y_train, X_train, fh=fh, **fit_params)
    111 
    112         else:  # if strategy == "update":

/databricks/python/lib/python3.8/site-packages/sktime/forecasting/base/_base.py in fit(self, y, X, fh)
    161         # we call the ordinary _fit if no looping/vectorization needed
    162         if not vectorization_needed:
--> 163             self._fit(y=y_inner, X=X_inner, fh=fh)
    164         else:
    165             # otherwise we call the vectorized version of fit

/databricks/python/lib/python3.8/site-packages/sktime/forecasting/compose/_pipeline.py in _fit(self, y, X, fh)
    243         name, forecaster = self.steps[-1]
    244         f = clone(forecaster)
--> 245         f.fit(y, X, fh)
    246         self.steps_[-1] = (name, f)
    247 

/databricks/python/lib/python3.8/site-packages/sktime/forecasting/base/_base.py in fit(self, y, X, fh)
    149 
    150         # check and convert X/y
--> 151         X_inner, y_inner = self._check_X_y(X=X, y=y)
    152 
    153         # set internal X/y to the new X/y

/databricks/python/lib/python3.8/site-packages/sktime/forecasting/base/_base.py in _check_X_y(self, X, y)
    975         if X is not None and y is not None:
    976             if self.get_tag("X-y-must-have-same-index"):
--> 977                 check_equal_time_index(X, y, mode="contains")
    978 
    979             if y_scitype != X_scitype:

/databricks/python/lib/python3.8/site-packages/sktime/utils/validation/series.py in check_equal_time_index(mode, *ys)
    272 
    273         if failure_cond:
--> 274             raise ValueError(msg)
    275 
    276 

ValueError: (time) indices of series 0 are not contained in index of series 0, extra indices are: PeriodIndex(['1947'], dtype='period[A-DEC]', freq='A-DEC')
```

**Comment by fkiraly** [comment]
Ok, I can confirm:

* under 0.11.2, this errors out with the separate bug https://github.com/alan-turing-institute/sktime/pull/2456
* on the fix branch #2456, this gives the same error, so: bug confirmed

**Comment by fkiraly** [comment]
reduced minimal example that produces the same error:

```python
from sktime.forecasting.compose import ForecastingPipeline 
from sktime.forecasting.fbprophet import Prophet
from sktime.transformations.series.difference import Differencer
from sktime.datasets import load_longley
y, X = load_longley()

#split train/test both y and X
fh = [1,2,3]
train_model, test_model = temporal_train_test_split(y, fh = fh)
X_train = X[X.index.isin(train_model.index)]
X_test = X[X.index.isin(test_model.index)]

#prophet pipeline
pipe_prophet = ForecastingPipeline(steps=[
        ("differencerX", Differencer(drop_na = True)),
        ("prophetforecaster", Prophet(yearly_seasonality='auto')),
        ]
)

#fit
pipe_prophet.fit(train_model, X = X_train, fh = [1,2,3])
```

**Comment by fkiraly** [comment]
this rules out problems with the inverse, since `ForecastingPipeline` does not call the differencer´s inverse

**Comment by fkiraly** [comment]
ok, I think this is not a bug, but expected behaviour.

Why:
* `Prophet` requires exogeneous data observed at all indices where a fit or forecast is being made
* `Differencer`, when called with the parameter `drop_na=True`, eats away the first index, hence making it unavailable for `Prophet`
* In the pipeline, the differencer eats one of the indices that `Prophet` would need, and the error message tells us that.

So, the solution would be to set `drop_na=False`.

For me, that leads to a different error, namely `Prophet` complaining about the index type. Which also seems reasonable to me, since the downstream `prophet` (from Facebook) doesn´t like the period index type of the Longley data (it's not an sktime restriction).

**Comment by fkiraly** [comment]
Thoughts, @anthonygiorgio97?

**Comment by anthonygiorgio97** [comment]
@fkiraly It's partially true, because I use `Differencer` both for exogenous X and endogenous y data. So the "double" `Differencer` operator would eat both the first index of y and the first index of X and both series should start at the second timepoint. Unless this kind of control is under `TransformedTargetForecaster`  and so before the `Differencer` operation on y.

If I use `drop_na = False` and I change index to timestamp, the prophet forecaster seems to initialize but then it stops giving me the error "Found NaN"

There is a way to perform the Differencer on both X and y series in a pipeline?

**Comment by fkiraly** [comment]
argh, ok. Yes, I see why you think the differencer eating both indices should work.

Unfortunately, it is not as straightforward, because you need to prevent any intermediate situation where an index in `y` is not there in `X`. My "naive" approach would be reversing the order of the pipelines, i.e., have the `TransformedTargetForecaster` *outside* and the `ForecastingPipeline` *inside*. That would eat away the first `y` index *first* and only *then* the first `X` index.

More generally, I feel it would be easier to deal with sth like this if the `Differencer` had a `fill_na` setting... perhaps a good idea for a new issue?

Regarding the "found NaN" problem, this is also coming from a property of `Prophet`, namely that it cannot deal with NaNs, I think? You could solve that by putting an `Imputer` in the pipeline before the NaNs reach `Prophet`.

Although the most elegant solution would be a `fill_na` in `Differencer`, that would allow to side-step both hacks which seem a bit excessive given that we are only dealing with a single nan here.

**Comment by anthonygiorgio97** [comment]
@fkiraly ok I have tried to reverse pipelines. Now if I use a `make_reduction` model like SVR() as follow, the pipeline work.

```
from sktime.transformations.series.boxcox import LogTransformer
from sktime.forecasting.compose import ForecastingPipeline, TransformedTargetForecaster
from sktime.transformations.series.compose import OptionalPassthrough
from sktime.forecasting.fbprophet import Prophet
from sktime.transformations.series.difference import Differencer
from sktime.forecasting.model_selection import ExpandingWindowSplitter, temporal_train_test_split, ForecastingGridSearchCV
from sktime.forecasting.compose import make_reduction
from sklearn.svm import SVR
from sktime.datasets import load_longley
y, X = load_longley()

#split train/test both y and X
fh = [1,2]
train_model, test_model = temporal_train_test_split(y, fh = fh)
X_train = X[X.index.isin(train_model.index)]
X_test = X[X.index.isin(test_model.index)]

#pipeline
pipe = TransformedTargetForecaster(steps = [
            ("log", OptionalPassthrough(LogTransformer())),
            ("differencer", Differencer(drop_na = True)),
            ("myforecasterpipe", ForecastingPipeline(steps = [
                ("logX", OptionalPassthrough(LogTransformer())),
                ("differencerX", Differencer(drop_na = True)),
                ("myforecaster", make_reduction(SVR(), window_length = 3))
        ]))  
])               

#cv setup
N_cv_fold = 1
step_cv = 1
cv = ExpandingWindowSplitter(initial_window = len(train_model) - (N_cv_fold - 1) * step_cv - len(fh),
                            start_with_window = True,
                            step_length = step_cv,
                            fh = fh)

param_grid = [{"log__passthrough": [False],
               "myforecasterpipe__logX__passthrough": [False]
              },
              {"log__passthrough": [True],
               "myforecasterpipe__logX__passthrough": [True]
              }
             ]

#grid search
gscv = ForecastingGridSearchCV(
            forecaster = pipe,
            cv=cv,
            param_grid=param_grid,
            verbose = 1
        )

#fit
gscv.fit(train_model)#, X = X_train)#, fh = fh)
```

However if I want to use Prophet now I face up with another error. This is the sample code I'm using:

```
from sktime.transformations.series.boxcox import LogTransformer
from sktime.forecasting.compose import ForecastingPipeline, TransformedTargetForecaster
from sktime.transformations.series.compose import OptionalPassthrough
from sktime.forecasting.fbprophet import Prophet
from sktime.transformations.series.difference import Differencer
from sktime.forecasting.model_selection import ExpandingWindowSplitter, temporal_train_test_split, ForecastingGridSearchCV
from sktime.datasets import load_longley
y, X = load_longley()

#split train/test both y and X
fh = [1,2]
train_model, test_model = temporal_train_test_split(y, fh = fh)
X_train = X[X.index.isin(train_model.index)]
X_test = X[X.index.isin(test_model.index)]

train_model.index = train_model.index.to_timestamp(freq='A')
X_train.index = X_train.index.to_timestamp(freq='A')
X_test.index = X_test.index.to_timestamp(freq='A')

#pipeline
pipe = TransformedTargetForecaster(steps = [
            ("log", OptionalPassthrough(LogTransformer())),
            ("differencer", Differencer(drop_na = True)),
            ("myforecasterpipe", ForecastingPipeline(steps = [
                ("logX", OptionalPassthrough(LogTransformer())),
                ("differencerX", Differencer(drop_na = True)),
                ("myforecaster", Prophet())
        ]))  
])               

#cv setup
N_cv_fold = 1
step_cv = 1
cv = ExpandingWindowSplitter(initial_window = len(train_model) - (N_cv_fold - 1) * step_cv - len(fh),
                            start_with_window = True,
                            step_length = step_cv,
                            fh = fh)

param_grid = [{"log__passthrough": [False],
               "myforecasterpipe__logX__passthrough": [False]
              },
              {"log__passthrough": [True],
               "myforecasterpipe__logX__passthrough": [True]
              }
             ]

#grid search
gscv = ForecastingGridSearchCV(
            forecaster = pipe,
            cv=cv,
            param_grid=param_grid,
            verbose = 1
        )

#fit
gscv.fit(train_model, X = X_train)#, fh = fh)
```

The error is this:

ValueError: Must supply freq for datetime value

It seems strange because freq is set as 'A' before fitting for train_model and X_train

**Comment by anthonygiorgio97** [comment]
Another strange behaviour is that if `fh=[1,2]` the code below works, but if `fh=[1,2,3]` or `[1,2,3,4],...` or longer, I have error:

```
from sktime.transformations.series.boxcox import LogTransformer
from sktime.forecasting.compose import ForecastingPipeline, TransformedTargetForecaster
from sktime.transformations.series.compose import OptionalPassthrough
from sktime.forecasting.fbprophet import Prophet
from sktime.transformations.series.difference import Differencer
from sktime.forecasting.model_selection import ExpandingWindowSplitter, temporal_train_test_split, ForecastingGridSearchCV
from sktime.forecasting.compose import make_reduction
from sklearn.svm import SVR
from sktime.datasets import load_longley
y, X = load_longley()

#split train/test both y and X
fh = [1,2,3]
train_model, test_model = temporal_train_test_split(y, fh = fh)
X_train = X[X.index.isin(train_model.index)]
X_test = X[X.index.isin(test_model.index)]

#pipeline
pipe = TransformedTargetForecaster(steps = [
            ("log", OptionalPassthrough(LogTransformer())),
            ("differencer", Differencer(drop_na = True)),
            ("scaler", TabularToSeriesAdaptor(StandardScaler())),
            ("myforecasterpipe", ForecastingPipeline(steps = [
                ("logX", OptionalPassthrough(LogTransformer())),
                ("differencerX", Differencer(drop_na = True)),
                ("scalerX", TabularToSeriesAdaptor(StandardScaler())),
                ("myforecaster", make_reduction(SVR()))
        ]))  
])

#cv setup
N_cv_fold = 1
step_cv = 1
cv = ExpandingWindowSplitter(initial_window = len(train_model) - (N_cv_fold - 1) * step_cv - len(fh),
                            start_with_window = True,
                            step_length = step_cv,
                            fh = fh)

param_grid = [{"log__passthrough": [False],
               "myforecasterpipe__logX__passthrough": [False],
               "myforecasterpipe__myforecaster__window_length": [2,3],
               "myforecasterpipe__myforecaster__estimator__C": [10,100]
              },
              {"log__passthrough": [True],
               "myforecasterpipe__logX__passthrough": [True],
               "myforecasterpipe__myforecaster__window_length": [2,3],
               "myforecasterpipe__myforecaster__estimator__C": [10,100]
              }
             ]

#grid search
gscv = ForecastingGridSearchCV(
            forecaster = pipe,
            cv=cv,
            param_grid=param_grid,
            verbose = 1
        )

#fit
gscv.fit(train_model, X = X_train)
```

Error:

`ValueError: could not broadcast input array from shape (5,2) into shape (1,5,3)`

I think the problem regards the `inverse_transform` of `Differencer` on exogenous variable used for validation, when we have to do prediction in validation set. But I don't understand why for forecasting horizon `[1,2]` it works.

**Comment by fkiraly** [comment]
> But I don't understand why for forecasting horizon [1,2] it works.

The explanation would be that the broadcast shape ends up being `(1, 5, 2)` which then works, by coincidence.

**Comment by fkiraly** [comment]
oh no 😢 

my suspicion about the two things:

* "must supply freq" is probably some transform eating away the freq. `pandas` indices are fiddly, and sometimes freq just disappears if you subset to an index of length 2 or less.
* the second one looks like some bad `numpy` call, perhaps `axis` is missing somewhere.

What would help is a "reduced" version of the above - would you be so kind to remove elements from the pipeline and imports until the error no longer occurs? I.e., try to construct the shortest example where we still have the problem.
I can then have a look at this during the week.

**Comment by fkiraly** [comment]
I went through the above and found that:

* the original bugs reported by @anthonygiorgio97, with the differencer not working, are resolved
* there are two bugs reported by @anthonygiorgio97 which were not resolved by the fixes and are still present in 0.12.0

I've now moved them to separate issues (#2806, #2807) so we can track better what is resolved and what is not.
